### PR TITLE
Add Metadata extraction from AnnData functionality (SCP-4824)

### PR DESCRIPTION
**Pull request by ehanna4**
Changes: 
- Alters the way we pass flags for file extraction from AnnData files in line with the updates done in ingest pipeline. This affects cluster and metadata extraction
- Update the ingest pipeline to be the most recent version

Additions:
- Logic for metadata extraction from AnnData files

This adds the Rails side for the ingest pipeline update you can find here: https://github.com/broadinstitute/scp-ingest-pipeline/pull/287

To test:
Pre-reqs
- You will need a clean study with no files in it
- You will need to make sure your local ingest-pipeline configuration is not set or is set to the latest release of 1.24.0.
- You need to ensure you've got the feature flag for ingesting anndata set to true for your user and study
- download this h5ad file (and unzip it, just couldn't put here without compressing it): 
[compliant_pbmc3K.h5ad.zip](https://github.com/broadinstitute/single_cell_portal_core/files/10362369/compliant_pbmc3K.h5ad.zip)

Steps:
- Pull this branch
- Start up a local instance and also start up delayed jobs
- Go to your clean study that has no files in it
- Go to file upload and navigate to the AnnData upload
- Upload this file:
- Once the file completes go to the study details page
- You should see:
Cluster info:
Screen Shot 2023-01-06 at 10 59 53 AM
Metadata info:
Screen Shot 2023-01-06 at 10 59 47 AM
And the AnnData file:
Screen Shot 2023-01-06 at 11 54 17 AM

You can also check out your studies google bucket to see the individual fragments:
Screen Shot 2023-01-06 at 11 29 25 AM

https://user-images.githubusercontent.com/54322292/211059786-f5fbb124-3235-4062-aa3c-82f3d88c99eb.mov

### CODE REVIEW

**Comment by ehanna4** [comment]
Adding test cases now

File: app/models/ingest_job.rb [hunk 1]
```diff
@@ -377,12 +379,13 @@ def set_study_state_after_ingest
     when :ingest_cell_metadata
       study.set_cell_count
       set_study_default_options
-      launch_subsample_jobs
+      launch_subsample_jobs unless study_file.is_anndata?
```
Comment (bistline) [inline app/models/ingest_job.rb:382]: Nice catch - this is one of the larger issues we have to tackle with AnnData ingest as we can't currently subsample with what we have

File: app/models/ingest_job.rb [hunk 2]
```diff
@@ -599,17 +602,29 @@ def set_anndata_file_info
 
   # launch appropriate downstream jobs once an AnnData file successfully extracts "fragment" files
   def launch_anndata_subparse_jobs
-    if params_object.extract_cluster.present?
+    
+    if params_object.attribute_as_array(:extract).include? 'cluster'
```
Comment (bistline) [inline app/models/ingest_job.rb:606]: I think might be easier to read as a case statement, especially once we add expression files to this.
```suggestion
params_object.attribute_as_array(:extract).each do |extract|
  case extract
  when 'cluster'
    params_object.attribute_as_array(:obsm_keys).each do |fragment|
      cluster_gs_url = params_object.fragment_file_gs_url(study.bucket_id, 'cluster', study_file.id, fragment)
      cluster_params = AnnDataIngestParameters.new(
        ingest_cluster: true, name: fragment, cluster_file: cluster_gs_url, domain_ranges: '{}',
        ingest_anndata: false, extract: nil, obsm_keys: nil, study_accession: nil,
      )
      job = IngestJob.new(study:, study_file:, user:, action: :ingest_cluster, params_object: cluster_params)
      job.delay.push_remote_and_launch_ingest
    end
  when 'metadata'
    metadata_gs_url = params_object.fragment_file_gs_url(study.bucket_id, 'metadata', study_file.id)
    metadata_params = AnnDataIngestParameters.new(
      ingest_cell_metadata: true, cell_metadata_file: metadata_gs_url,
      ingest_anndata: false, extract: nil, obsm_keys: nil, study_accession: study.accession,
    )
    job = IngestJob.new(study:, study_file:, user:, action: :ingest_cell_metadata, params_object: metadata_params)
    job.delay.push_remote_and_launch_ingest
  end
end
```

File: app/models/ann_data_ingest_parameters.rb [hunk 3]
```diff
@@ -49,13 +61,26 @@ def initialize(attributes = {})
...
-      ingest_anndata:, anndata_file:, extract_cluster:, obsm_keys:, ingest_cluster:, cluster_file:, name:, domain_ranges:
+      ingest_anndata:, anndata_file:, extract:, obsm_keys:, ingest_cluster:, cluster_file:,
+      name:, domain_ranges:, cell_metadata_file:, ingest_cell_metadata:, study_accession:
     }.with_indifferent_access
   end
 
   # generate a GS URL to a derived fragment that was extracted from the parent AnnData file
-  def fragment_file_gs_url(bucket_id, fragment_type, name)
-    "gs://#{bucket_id}/_scp_internal/anndata_ingest/#{name}.#{fragment_type}.anndata_segment.tsv"
+  # File name structure is: <input_filetype>_frag.<file_type>.<file_type_detail>.tsv
```
Comment (bistline) [inline app/models/ann_data_ingest_parameters.rb:70]: Nice comments, and I like how this is being structured.

**Review by bistline — APPROVED** [review]
Looks good, and works as advertised!  One small non-blocking suggestion.

**Review by eweitz — APPROVED** [review]
Code and demo look good!  This would be a cool topic for demo.

Non-blocking: the test steps say "Upload this file:", without indicating the file.  I imagine they should say something like "Upload compliant_pbmc3K.h5ad, which you decompressed per above prerequisites".

**Pull request merged**